### PR TITLE
Add tx weighting stage

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Transaction don't have the stake weight of the sender and to order them it would be nice to have.

#### Summary of Changes

Add a stage to add the stake weights after streamer and before sigverify.

Fixes #

### CODE REVIEW

**Comment by buffalu** [comment]
do top validators know this is coming? hopefully they’re running more than 1G NICs lol

**Comment by buffalu** [comment]
also this could make DOS worse without special forwarder software?

not only can you screw with block propagation of current leader by stealing bandwidth from them, but also can forward my DOS packets through the highest staker to the current leader

**Comment by sakridge** [comment]
> also this could make DOS worse without special forwarder software?
> 
> not only can you screw with block propagation of current leader by stealing bandwidth from them, but also can forward my DOS packets through the highest staker to the current leader

That high staker which is forwarding will QoS those packets and will prefer to send packets from other staked nodes first before it sends the unstaked DoS packets.

File: core/src/transaction_weighting_stage.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                        ips = staked_nodes
+                            .iter()
+                            .filter_map(|(pubkey, stake)| {
+                                cluster_info
+                                    .lookup_contact_info(pubkey, |contact_info| {
+                                        contact_info.tvu.ip()
+                                    })
+                                    .map(|ip| (ip, *stake))
+                            })
+                            .collect();
```
[the remark this answers is not in the document]
Comment (sakridge) [reply core/src/transaction_weighting_stage.rs]: yea, thanks, changed to that

**Comment by segfaultdoc** [comment]
> > also this could make DOS worse without special forwarder software?
> > not only can you screw with block propagation of current leader by stealing bandwidth from them, but also can forward my DOS packets through the highest staker to the current leader
> 
> That high staker which is forwarding will QoS those packets and will prefer to send packets from other staked nodes first before it sends the unstaked DoS packets.


> #### Summary of Changes

Should those changes to the QoS service be added to this PR?

File: core/src/transaction_weighting_stage.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                }
+            })
+            .unwrap();
+        Self { thread_hdl }
+    }
+
+    fn apply_weights(batches: &mut [PacketBatch], ip_to_stake: &HashMap<IpAddr, u64>) {
+        batches.into_par_iter().for_each(|batch| {
+            batch.packets.par_iter_mut().for_each(|packet| {
+                packet.meta.weight = *ip_to_stake.get(&packet.meta.addr().ip()).unwrap_or(&0);
```
Comment (taozhu-chicago) [inline core/src/transaction_weighting_stage.rs:69]: why need to call `.ip()` on `packet.meta.addr()`, which is `IpAddr`.

**Comment by michaelh-laine** [comment]
How do tx first get into "the system" as most non-vote tx are first submitted to/via RPC nodes which have zero stake. Does this PR not create a scenario where votes push out tx from all other clients? And if there is a sequence of low-staked leaders that receive a burst of client transactions (as RPC forwards to upcoming leaders) they then receive low priority when forwarded (if not immediately included in a block). 

Perhaps there's more intelligence in the mechanism that would address these issues, would appreciate some clarity on this.

(Question derived from discussion on Validator roundtable call on 20.01.2022)

**Comment by taozhu-chicago** [comment]
> How do tx first get into "the system" as most non-vote tx are first submitted to/via RPC nodes which have zero stake. Does this PR not create a scenario where votes push out tx from all other clients? 
In short, no, votes will not push out low-staked non-vote transactions. 
Votes and non-vote transactions are in separate leader threads, with separate queues. The stake weighted shuffle applies to each queue independently.

**Comment by taozhu-chicago** [comment]
Close this PR as it has been merged into #22958, reviews and comments should be made on combined PR.

**Pull request closed without merging**